extra_data/auxiliary.py

```diff
@@ -0,0 +1,155 @@
+            dc = H5File(fa.filename)
+
+            source_file_paths = set()
+            for source in dc.all_sources:
+                sd = dc[source]
+                source_file_paths |= set(sd[sd.one_key()].source_file_paths)
+
+            for file in source_file_paths:
+                new_files.append(FileAccess(file))
+
+        return new_files
+
+    @property
+    def reduction_sources(self):
+        return frozenset([src for src, sd in self._sources_data.items()
+                          if sd.is_reduction])
+
+    @property
+    def errata_sources(self):
+        return frozenset([src for src, sd in self._sources_data.items()
+                          if sd.is_errata])
+
+    @property
+    def all_sources(self):
+        return frozenset(self._sources_data.keys())
+
+    def __contains__(self, item):
+        return DataCollection.__contains__(self, item)
+
+    def __str__(self):
+        num_trains = len(next(iter(self._sources_data.values())).train_ids) \
+            if self._sources_data else 0
+
+        return f'<extra_data.AuxiliaryIndexer for {len(self.all_sources)} ' \
+               f'sources and {num_trains} trains>'
+
+    def _source_info(self, label, sources, details_for_sources=(),
+                     with_aggregators=False):
+        details_sources_re = [re.compile(fnmatch.translate(p))
+                              for p in details_for_sources]
+        info_str = ''
+
+        def src_data_detail(s, keys, prefix=''):
+            """Detail for how much data is present for an instrument group"""
+            if not keys:
+                return
+
+            sd = self[s]
+            counts = sd[list(keys)[0]].data_counts()
+            ntrains_data = (counts > 0).sum()
+            return f'{prefix}data for {ntrains_data} trains ' \
+                   f'({ntrains_data / len(sd.train_ids):.2%}), ' \
+                   f'up to {counts.max()} entries per train\n'
+
+        def keys_detail(s, keys, prefix=''):
+            """Detail for a group of keys"""
+            for k in keys:
+                kd = self[s, k]
+                entry_info = f', entry shape {kd.entry_shape}'  \
+                    if kd.entry_shape else ''
+
+                return f'{prefix}{k}\t[{kd.dtype}{entry_info}]\n'
+
+        info_str += f'{len(sources)} {label} sources:\n'
+
+        for s in sorted(sources):
+            agg_str = f' [{self[s].aggregator}]' if with_aggregators else ''
+            info_str += f'  - {agg_str} {s}\n'
+            if not any(p.match(s) for p in details_sources_re):
+                continue
+
+            # Detail for instrument sources:
+            for group, keys in groupby(sorted(self[s].keys()),
+                                       key=lambda k: k.split('.')[0]):
+                info_str += f'    - {group}:\n'
+                keys = list(keys)
+                info_str += src_data_detail(s, keys, prefix='      ')
+                info_str += keys_detail(s, keys, prefix='      - ')
```

Comment:
Building up a long string by concatenation has poor performance. It's *probably* not big enough to matter in this case, but still. The better options are to add pieces to a list and use `''.join(l)`, or write into a StringIO object.

https://docs.python.org/3/library/stdtypes.html#:~:text=Concatenating%20immutable%20sequences